cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/common/MetadataClient.java

```diff
@@ -4,123 +4,164 @@
         }
-        if (MonitorUtils.metadataChanged(beforeUpdate, _metadata.fetch())) {
+        Cluster newCluster = refreshedCluster.get();
+        if (MonitorUtils.metadataChanged(_cluster, newCluster)) {
           _metadataGeneration.incrementAndGet();
+          _cluster = newCluster;
         }
       } else {
         LOG.warn("Failed to update metadata in {}ms. Using old metadata with version {} and last successful update {}.",
-                 timeout, _metadata.version(), _metadata.lastSuccessfulUpdate());
+                 timeout, _version, _lastSuccessfulUpdate);
       }
     }
-    return new ClusterAndGeneration(_metadata.fetch(), _metadataGeneration.get());
+    return new ClusterAndGeneration(_cluster, _metadataGeneration.get());
   }
 
-  /**
-   * Close the metadata client. Synchronized to avoid interrupting the network client during a poll.
-   */
-  public synchronized void close() {
-    _networkClient.close();
+  private Cluster doRefreshMetadata(int timeout) throws InterruptedException, ExecutionException, TimeoutException {
+    // We use both AdminClient timeout options and `future.get(timeout)`
+    // The former ensures the timeouts are enforced in the brokers
+    int remaining = timeout;
+    long start = _time.milliseconds();
+
+    Set<String> topicNames = _adminClient.listTopics(new ListTopicsOptions().timeoutMs(remaining))
+        .names().get(remaining, TimeUnit.MILLISECONDS);
+    long end = _time.milliseconds();
+    long elapsedTimeMs = end - start;
+    remaining -= elapsedTimeMs;
+
+    DescribeClusterResult result = _adminClient.describeCluster(new DescribeClusterOptions().timeoutMs(remaining));
+    Collection<Node> nodes = result.nodes().get(remaining, TimeUnit.MILLISECONDS);
+    String clusterId = result.clusterId().get(remaining, TimeUnit.MILLISECONDS);
+    Node controller = result.controller().get(remaining, TimeUnit.MILLISECONDS);
+    elapsedTimeMs = _time.milliseconds() - end;
+    remaining -= elapsedTimeMs;
+
+    Map<String, TopicDescription> topicDescriptions = _adminClient.describeTopics(topicNames, new DescribeTopicsOptions().timeoutMs(remaining))
+        .all().get(remaining, TimeUnit.MILLISECONDS);
+
+    return cluster(clusterId, nodes, controller, topicDescriptions);
+  }
+
+  private Cluster cluster(String clusterId, Collection<Node> nodes, Node controller,
+                          Map<String, TopicDescription> describeTopicResult) {
+    List<PartitionInfo> partitionInfos = new LinkedList<>();
+    Set<String> internalTopics = new HashSet<>();
+
+    for (Map.Entry<String, TopicDescription> topicDescription: describeTopicResult.entrySet()) {
+      if (topicDescription.getValue().isInternal()) {
+        internalTopics.add(topicDescription.getKey());
+      }
+
+      for (TopicPartitionInfo topicPartitionInfo : topicDescription.getValue().partitions()) {
+        PartitionInfo partitionInfo = new PartitionInfo(topicDescription.getKey(),
```

Comment:
How does this call capture offline replicas?

Reply:
`topicPartitionInfo.replicas()` would return any replicas that are offline, unless you are talking about something else

Reply:
Good catch